decision-records/2024-08-29--dashboard-uid-collision.md

```diff
@@ -0,0 +1,134 @@
+3. UID length must be <= 40 (grafana limit).
+4. Should be able to easily tell from which charm each dashboard came.
+
+## Decision
+(2/P) + (4/R) + (5/P) + (6).
+
+## Alternatives considered
+Deduplication can be addressed on the provider (charm with dashboards) or the requrier side (grafana charm).
+In the following subsections, `P` and `R` stand for provider and requirer.
+
+### (1/P) Use a prescribed naming pattern for UIDs at dev time
+We can update the [best practices doc](https://discourse.charmhub.io/t/grafana-k8s-docs-how-to-create-a-great-charmed-dashboard/14188#use-an-effective-naming-scheme-22),
+and it's up to charm authors to be disciplined about adhering to it. Charm lib will raise if the pattern doesn't match the expected ("test-time linter").
+
+A good option for UIDs could be `f"{charm_name}-{codename}"`, where `codename` is e.g. an abbreviate dashboard title that the charm author needs to make sure is unique within the charm.
+
+Benefits:
+- UID is human-readable and sensible.
+
+Disadvantages:
+- Coming up with a codename is toilsome.
+- To adhere to this new standard, would need to modify all existing dashboards.
+
+### (2/P) Automatically overwrite UIDs with a prescribed naming pattern at deploy/relate time, on the provider side
+SHA1 is of length 40 (`len(hashlib.sha1(b"whatever").hexdigest()) == 40`), so calculate the dashboards UID from hashing (charm_name + relative_path of the dashboard file).
+
+Benefits:
+- The combination of charm_name and relative_path is almost guaranteed to be unique across the ecosystem. Exceptions:
+  - Problem (a)
+  - Local charm's name and dashboard relative_path collide with another charm.
+- Change will be automatically rolled out to everyone by updating the dashboards charm lib.
+
+Disadvantages:
+- Keeping the dashboard itself intact but moving it to another subfolder would break dashboard URL.
+- May need special attention for aggregation (grafana-agent, cos-proxy) and side-loading (cos-configuration).
+- Overwriting without asking might be surprising for users who want to control UIDs themselves.
+- Dashboards taken from the grafana marketplace would lose their original identity.
+
+### (3/R) Automatically overwrite UIDs with a prescribed naming pattern on relation changed, on the requirer side
+This is possible, thanks to the schema we already have in place (see Appendix 1).
+
+Similar benefits and disadvantages to doing it on the provider side.
+
+### (4/R) On rev-collision, keep only the latest dashboard revision
+
+If we received a "newer dashboard" from a higher revision of a charm, we'd only keep the latest dashboard.
+
+Benefits:
+- A workaround for problem 1a.
```

Comment:
Is problem 1a: "`a) Two charms on different revisions`"?

Reply:
Yes. Clarified.